teuthology/scrape.py

```diff
@@ -0,0 +1,519 @@
+            if re.match(regex, job.get_failure_reason()) is not None:
+                return True
+
+        return False
+
+
+class AssertionReason(Reason):
+    def __init__(self, job):
+        self.assertion = job.get_assertion()
+        self.backtrace = job.get_backtrace()
+
+    def get_description(self):
+        return "Assertion: {0}".format(self.assertion)
+
+    def get_detail(self):
+        return self.backtrace
+
+    @classmethod
+    def could_be(cls, job):
+        return job.get_assertion() is not None
+
+    def match(self, job):
+        return self.assertion == job.get_assertion()
+
+
+class LockdepReason(AssertionReason):
+    """
+    Different to a normal assertion, because matches should not only
+    have the same assertion but the same backtrace (don't want to glob
+    all lockdep failures together if they are really being tripped in
+    different places)
+    """
+    @classmethod
+    def could_be(cls, job):
+        if not super(LockdepReason, cls).could_be(job):
+            return False
+
+        if "common/lockdep" in job.get_assertion():
+            return True
+        else:
+            return False
+
+    def get_description(self):
+        return "Lockdep: {0}".format(self.assertion)
+
+    def match(self, job):
+        if not super(LockdepReason, self).match(job):
+            return False
+
+        if self.backtrace:
+            if job.get_backtrace():
+                ratio = difflib.SequenceMatcher(None, self.backtrace, job.get_backtrace()).ratio()
+                return ratio > 0.5
+            else:
+                return False
+        else:
+            # No backtrace to compare about, allow matches based purely on assertion
+            return True
+
+
+class DeadReason(Reason):
+    """
+    A reason for picking up jobs with no summary.yaml
+    """
+    def __init__(self, job):
+        self.description = "Dead"
+        self.last_tlog_line = job.get_last_tlog_line()
+        self.backtrace = job.get_backtrace()
+
+    def get_description(self):
+        return "Dead: {0}".format(self.last_tlog_line)
+
+    def get_detail(self):
+        return self.backtrace
+
+    @classmethod
+    def could_be(cls, job):
+        return job.summary_data is None
+
+    def match(self, job):
+        if job.summary_data is not None:
+            return False
+
+        if self.backtrace:
+            if job.get_backtrace():
+                # We both have backtrace: use that to decide if we're the same
+                ratio = difflib.SequenceMatcher(None, self.backtrace, job.get_backtrace()).ratio()
+                return ratio > 0.5
+            else:
+                # I have BT but he doesn't, so we're different
+                return False
+
+        if self.last_tlog_line is None and job.get_last_tlog_line() is None:
+            return True
+
+        ratio = difflib.SequenceMatcher(None, self.last_tlog_line, job.get_last_tlog_line()).ratio()
+        return ratio > 0.5
```

Comment:
For better readability this can be rewritten like:
```
if self.last_tlog_line or job.get_last_tlog_line():
          ratio = difflib.SequenceMatcher(None, self.last_tlog_line,
                                          job.get_last_tlog_line()).ratio()
          return ratio > 0.5
return True
```